commands/provider/marketplace_provider.go

```diff
@@ -18,46 +19,67 @@ func NewMarketplaceProvider(c coreapi.CoreClient) *MarketplaceProvider {
 	return &MarketplaceProvider{client: client{c}}
 }
 
-// PublishServiceVersion executes the create service version task
-func (p *MarketplaceProvider) PublishServiceVersion(service MarketplaceManifestServiceData, from string) (Transaction, error) {
-	input := marketplacePublishServiceVersionTaskInputs{
-		marketplaceTransactionTaskInputs: marketplaceTransactionTaskInputs{From: from},
-		Service:                          service,
+// PreparePublishServiceVersion executes the create service version task
+func (p *MarketplaceProvider) PreparePublishServiceVersion(service MarketplaceManifestServiceData, from string) (Transaction, error) {
+	input := marketplacePreparePublishServiceVersionTaskInputs{
+		marketplacePrepareTaskInputs: marketplacePrepareTaskInputs{From: from},
+		Service:                      service,
 	}
 	var output Transaction
-	return output, p.call("publishServiceVersion", input, &output)
+	return output, p.call("preparePublishServiceVersion", input, &output)
 }
 
-// CreateServiceOffer executes the create service offer task
-func (p *MarketplaceProvider) CreateServiceOffer(sid string, price string, duration string, from string) (Transaction, error) {
-	input := marketplaceCreateServiceOfferTaskInputs{
-		marketplaceTransactionTaskInputs: marketplaceTransactionTaskInputs{From: from},
-		Sid:                              sid,
-		Price:                            price,
-		Duration:                         duration,
+// PublishPublishServiceVersion executes the task publish service version task
+func (p *MarketplaceProvider) PublishPublishServiceVersion(signedTransaction string) (sid, versionHash, manifest, manifestProtocol string, err error) {
+	input := marketplacePublishTaskInputs{
+		SignedTransaction: signedTransaction,
 	}
-	var output Transaction
-	return output, p.call("createServiceOffer", input, &output)
+	var o marketplacePublishPublishServiceVersionTaskOutputs
+	err = p.call("publishPublishServiceVersion", input, &o)
+	return o.Sid, o.VersionHash, o.Manifest, o.ManifestProtocol, err
```

Comment:
Why not keeping the same pattern in the code. You could expose `marketplacePublishPublishServiceVersionTaskOutputs` and have something like that
```go
var o marketplacePublishPublishServiceVersionTaskOutputs
return o, p.call("publishPublishServiceVersion", input, &o)
```
I feel a function that returns too much data should return a struct.

Same for the `PublishCreateServiceOffer` and `PublishPurchase`

Reply:
Yes, the function has too many returned value but:
- the CLI doesn't directly depend on the package provider, all the function are declared in commands.go
- the commands actually only read 2 or 3 returned values. Should I only returned value useful for the commands?

Reply:
> the CLI doesn't directly depend on the package provider, all the function are declared in commands.go

It does on few functions already: `UploadSource`, `PreparePublishServiceVersion`, `PrepareCreateServiceOffer`, `PreparePurchase`, `GetService`... basically all except these ones

> it returns only simple types

No it doesn't:
```
PreparePublishServiceVersion(service provider.MarketplaceManifestServiceData, from string) (provider.Transaction, error)
GetService(sid string) (provider.MarketplaceService, error)
```

> the commands actually only read 2 or 3 returned values. Should I only returned value useful for the commands?
 
If we need only these data we could just return them but if we have the struct let's return the struct like that we have all the data without having a long list of outputs